### DISCUSSION TITLE
LUA APIs - LUA Documentation - JSON parsing - URL protocols - Official Discord Server - Donations

### DISCUSSION THREAD

**Discussion by shroomjaks** [YimMenu/YimMenu#1884] [Ideas]
I suggest a couple LUA apis to be added as they could create an awesome ecosystem of YimMenu scripts that no other GTA mod menu would have.

I would suggest taking notes from a popular program called Synapse X which is a Roblox LUA executor, features such as:
- https://docs.synapse.to/development/web_api.html - HTTP API.
- https://docs.synapse.to/development/filesystem_api.html - Safe filesystem API, only allows data to be saved to a ./workspace folder, blocks scripts from saving executables and also blocks scripts from creating files in a loop to take up system storage.
- https://docs.synapse.to/reference/websocket_lib.html - Websocket API.
- https://docs.synapse.to/reference/crypt_lib.html - Crypto API.
- https://docs.synapse.to/reference/input.html - Mouse and keyboard input API, inputs don't work outside of the game window.
- JSON stringifying and parsing API, like JavaScript

I think this GitHub repo contains Synapse X source code from a few years back, it should have most of the features listed above, but I'm not sure.
https://github.com/Acrillis/SynapseX

Also a proper YimMenu LUA documentation page is needed.

Examples of what could be done if these features were implemented:
1. A GitHub repository containing a JSON file of known chat spamming bots, the LUA script will fetch the database on startup, parse the database, parse your players.json file and add any chat spamming bots that aren't already in your players.json and enable the block_join value for them, effectively banning these bots from every session with a user who is running YimMenu and has this script.
2. A websocket server that allows YimMenu users to communicate through a global YimMenu chat
3. Custom macros through the input API
4. Connecting to a Node.js program running locally on the users computer, through either a websocket or HTTP post requests to create a custom Discord RPC status, could show details like if the user is playing GTA in single player or multiplayer, the amount of players they're playing with, etc. (Discord RPC could even be added as a native opt-in feature to YimMenu and customizable through a LUA script and the YimMenu UI, this could also serve as free advertisement for YimMenu)

Another feature I would suggest is a way for YimMenu to have it's own URL protocols to launch GTA, connect to sessions, join people via their RID, etc. Such as:
`yimmenu://launch`
`yimmenu://launch/joinplayer/{RID}`
This would require YimMenu to be an EXE file rather than a DLL and to have it's own service running in the background.

The features I'm suggesting are super ambitious but could turn YimMenu into the best free mod menu ever through the community and the ease of access of the LUA scripting language.

Lastly, YimMenu needs an official Discord server! This way users can get announcements about updates to YimMenu, users can share their LUA scripts, users can chat and suggest their ideas easily, and contributors / developers can communicate together to add new features faster, and some way to be able to donate to YimMenu (I just remembered Discord Server Subscriptions are a thing, donations could be done that way)

Also, another thing that would be awesome is a way to use LuaRocks packages (if possible) within the LUA scripts. Not sure how this would work and if it even would.
https://luarocks.org

**Comment by tupoy-ya** [comment]
> Lastly, YimMenu needs an official Discord server! This way users can get announcements about updates to YimMenu, users can share their LUA scripts, users can chat and suggest their ideas easily, and contributors / developers can communicate together to add new features faster, and some way to be able to donate to YimMenu.

Discard server got banned, and matrix works good enough in my opinion. 
https://github.com/YimMenu/YimMenu/discussions/361#discussioncomment-3732080

As for a place to develop scripts: https://github.com/YimMenu-Lua

**Reply by shroomjaks** [reply]
Thanks for the information, does Matrix support bots? If not, I think something like Guilded, Telegram or Signal would be better. I'm leaning more towards Guilded as it's the most similar to Discord, supports bots, and I've seen other mod menus use it for their communities.

**Reply by tupoy-ya** [reply]
It does support bots, we had a discord bridge bot before discord got banned.

I'm personally against Guilded because of it's proprietary license.
Telegram or signal chats can be made, plus they can be bridged to matrix just like discord was.

**Reply by Yimura** [reply]
Any reason for the need of a bot?

Also we can never ever have any bots again in the Matrix due to the encrypted nature of the channels.

**Reply by tupoy-ya** [reply]
> Also we can never ever have any bots again in the Matrix due to the encrypted nature of the channels.

Pretty sure you can make a bot that works in encrypted rooms, it's just harder then in unencrypted ones.
Because as far as i know there is no difference between users and bots in matrix.

**Reply by Yimura** [reply]
I guess that's true, we'll probably be stuck with having to create one by ourselves though as the Discord bridge for example just couldn't deal with encrypted rooms.

**Reply by tupoy-ya** [reply]
> I guess that's true, we'll probably be stuck with having to create one by ourselves though as the Discord bridge for example just couldn't deal with encrypted rooms.

From what i have read in their blog post about encryption they won't be able handle the amount of traffic going through the bridges if they enable encryption.
Some of their other bots do work with encryption however.

> Most notably, [Voyager](https://t2bot.io/voyager), the [Discord](https://t2bot.io/discord) and [Telegram](https://t2bot.io/telegram) bridges, and several other integrations are not listed here. Voyager isn’t included because the bot needs further updates to handle the private data contained in encrypted rooms. The bridges are excluded because they aren’t able to handle the traffic at the scale needed by t2bot.io - improvements are expected in this area over time, however.
https://t2bot.io/blog/2020/10/30/enabling-encryption-for-bots/

**Comment by Yimura** [comment]
All these additions seem fun but could pose a serious security threat.

### Web requests

Can be used to exfiltrate data, from the games memory space like email address and other personal identifiable information, combined with the other features you suggested could form a serious threat.

### File System API

No need for a special package, we have a File Manager service in YimMenu we just need to guard it against directory traversal attacks.

### WebSocket

Probably shouldn't be accessible from Lua, if anything should be directly implemented in YimMenu.
And as previously stated there are several reason on why we will never add a YimMenu Global Chat (see #1222).

### Crypto API

I don't immediately see the use of this but I'll be happy to hear your opinion.

### Input API

The natives from the game for controls are exposed these can be used to do actions in Lua.
And for general purpose inputs we've exposed everything from ImGui so you can use these to get any other sort of inputs you may need.

### JSON

Sure this sounds like a valid request.

### Discord RPC

Sure this sounds like a fun feature to add but only directly in the menu and not through Lua.

### Protocol Handlers

`yimmenu://`

Sure that sounds like something fun to implement and could possibly be implemented directly into the DLL but should probably be implemented into a YimMenu launcher at some point.

### Lua Rocks

Sounds like fun but is probably another security nightmare, you've got to understand that users don't care about any warnings or popups.
They'll just sheepishly accept it and then have all their tokens, passwords stolen.

### Lua sharing

Go here to request Lua repository: https://github.com/YimMenu-Lua/submission/issues/new/choose
In the future we'll make an interface to interact with these repositories and download these to your YimMenu instance directly.

https://github.com/orgs/YimMenu-Lua/repositories
For now any scripts grabbed from there will need to be manually put into the `%appdata%/scripts/` folder.

**Reply by shroomjaks** [reply]
>Web requests
Can be used to exfiltrate data, from the games memory space like email address and other personal identifiable information, combined with the other features you suggested could form a serious threat.

What about having the feature be able to be enabled through a developer mode option? Something like where you would have to open up a file in the YimMenu source and edit the code to enable it, next to the variable that enables the feature there should be a comment explaining the dangers of enabling HTTP requests, then they can build the YimMenu DLL with the option enabled. This way it can sort of prove that the user knows what they're doing when enabling HTTP requests. I would love to have it as a personal feature, like sending Discord webhook requests and querying my own databases and such, and also having the same option with websockets. Everything else you mentioned makes sense to me.

Similar to how Discord warns you when you open the developer console:
![image](https://github.com/YimMenu/YimMenu/assets/46830521/cf503c7b-c1ae-474a-83c2-628ce911900a)

**Reply by Harmless05** [reply]
Web requests could be used to update one's Lua script. Or once there is a UI to download scripts for YimMenu, we could also add a Lua scripts updater.

**Reply by Yimura** [reply]
No even then, in combination with people publishing their scripts on the YimMenu-Lua organisation and then making releases on there people could push updates for their scripts.
https://github.com/orgs/YimMenu-Lua/repositories

If people have issues with scripts they can go directly to the Github repo of the script and open an issue to state the problems they are experiencing.

All of this should be handled directly from the YimMenu dll and not be exposed to the Lua scripts directly.

**Reply by Harmless05** [reply]
Yeah, true.
I'm currently keeping my basic Lua scripts under my account repo because I'm too lazy to move everything to YimMenu-Lua

**Reply by Harmless05** [reply]
Also, might I say that this GitHub Discussions feature is amazing. If the community wanted, there is even an option to do polls on what users actually want in the menu.
